### PR TITLE
Handle common cases in arg validation

### PR DESCRIPTION
**Pull request by tyler-ball**
```
[clibuddy]$ be bin/clibuddy run chef-do --help=arg
Could not parse the arguments provided to 'chef-do': needless argument: --help=arg
```

### CODE REVIEW

File: lib/clibuddy/runner.rb [hunk 1]
```diff
@@ -159,7 +159,11 @@ def lookup_command(name, provided_args)
         # A: we can do differently-formatted exception handling that makes it clear.
         raise Errors::NoSuchCommand.new(name)
       end
-      InterpretedCommand.new(cmd, provided_args)
+      begin
+        InterpretedCommand.new(cmd, provided_args)
+      rescue OptionParser::ParseError => e
+        raise Errors::OptionParserError.new(cmd.name)
```
Comment (marcparadise) [inline lib/clibuddy/runner.rb]: Is there any value in the original message from ParseError?

File: lib/clibuddy/interpreted_command.rb [hunk 2]
```diff
@@ -13,24 +14,57 @@ class InterpretedCommand
...
+    # easily returned from #mapped_args later
+    def validate_args
+      opt_parser = OptionParser.new do |opts|
+        cmd.definition.flags.each do |flag|
+          parser_args = []
+          l = "--#{flag.flag}"
+          if flag.arg
+            if flag.arg =~ /^\[(.+)+\]$/
+              # if the arg is optional (surrounded by []) then the long form must look like `--key[=VALUE]` while
+              # the short form must look like `-k[VALUE]`
```
Comment (marcparadise) [inline lib/clibuddy/interpreted_command.rb:49]: Eventually we'll want to revisit this --  I'm a firm believer in  "--x=VAL" vs "--x VAL" being equally acceptable (though I know OptParser doesn't agree :D)

File: lib/clibuddy/main.rb [hunk 3]
```diff
@@ -66,6 +66,7 @@ def cli_buddy_error(e)
       t << [p.bold("Error Code "), e.id]
       t << [p.bold("File "), @descriptor_file]
       t << [p.bold("Message "), e.message]
+      t << [p.bold("Cause "), e.cause] if e.cause
```
Comment (marcparadise) [inline lib/clibuddy/main.rb]: What kind of info will this usually be?

**Review by marcparadise — APPROVED** [review]
Looks good! Couple of minor non-blocking questions inline.

File: lib/clibuddy/runner.rb (hunk 1 above)
Comment (tyler-ball) [reply lib/clibuddy/runner.rb]: Yes, and we get that with the `e.cause` logic I added into the error printer

File: lib/clibuddy/interpreted_command.rb (hunk 2 above)
Comment (tyler-ball) [reply lib/clibuddy/interpreted_command.rb:49]: Agreed, we should revisit, but I don't think we're using flags for any part of our flow now so I didn't see it get angry at me

File: lib/clibuddy/main.rb (hunk 3 above)
Comment (tyler-ball) [reply lib/clibuddy/main.rb]: I only saw it being used for the ParseError messages and those generally were very helpful. But we could always restrict this later if we see it being not useful.

**Pull request merged**